source/MapPanel.cpp

```diff
@@ -1059,32 +1070,35 @@ void MapPanel::DrawWormholes()
 			continue;
 
 		for(auto &&link : it.second.Links())
-			if(p.IsInSystem(link.first)
-					&& player.HasVisited(*link.first) && player.HasVisited(*link.second))
-				arrowsToDraw.emplace(link.first, link.second);
+			if(p.IsInSystem(link.first) && player.HasVisited(*link.first) && player.HasVisited(*link.second))
+				arrowsToDraw.emplace_back(link.first, link.second, it.second.GetLinkColor());
 	}
 
-	const Color &wormholeDim = *GameData::Colors().Get("map unused wormhole");
-	const Color &arrowColor = *GameData::Colors().Get("map used wormhole");
 	static const double ARROW_LENGTH = 4.;
 	static const double ARROW_RATIO = .3;
 	static const Angle LEFT(30.);
 	static const Angle RIGHT(-30.);
 	const double zoom = Zoom();
 
-	for(const pair<const System *, const System *> &link : arrowsToDraw)
+	for(const WormholeArrow &link : arrowsToDraw)
 	{
+		// Get the wormhole link color.
+		const Color &arrowColor = *link.color;
+		const Color &wormholeDim = Color::Combine(1.f, arrowColor, -0.66f, arrowColor);
+
 		// Compute the start and end positions of the wormhole link.
-		Point from = zoom * (link.first->Position() + center);
-		Point to = zoom * (link.second->Position() + center);
+		Point from = zoom * (link.from->Position() + center);
+		Point to = zoom * (link.to->Position() + center);
 		Point offset = (from - to).Unit() * LINK_OFFSET;
 		from -= offset;
 		to += offset;
 
 		// If an arrow is being drawn, the link will always be drawn too. Draw
 		// the link only for the first instance of it in this set.
-		if(link.first < link.second || !arrowsToDraw.count(make_pair(link.second, link.first)))
-			LineShader::Draw(from, to, LINK_WIDTH, wormholeDim);
+		if(link.from < link.to || count_if(arrowsToDraw.begin(), arrowsToDraw.end(),
+			[link](const WormholeArrow &cmp)
```

Comment:
This should perhaps capture the link as a reference?

Reply:
```suggestion
			[&link](const WormholeArrow &cmp)
```

I did this for Azure XD

Reply:
wait... is this correct?

Reply:
yea, should be fine